app/controllers/document_images_controller.rb

```diff
@@ -0,0 +1,168 @@
+# frozen_string_literal: true
+
+require 'mini_magick'
+
+class DocumentImagesController < ApplicationController
+  def edit
+    @document = Document.find(params[:id])
+  end
+
+  def update
+    document = Document.find(params[:id])
+    upload = UploadedImageProcessorService.new(image_params[:image]).process
+
+    if !upload.valid?
+      redirect_to document_images_path(document),
+                   flash: { notice: "Image upload failed", errors: upload.errors }
+    else
+      blob = ActiveStorage::Blob.build_after_upload(
+        io: upload.file,
+        content_type: upload.mime_type,
+        filename: upload.filename,
+      )
+      blob.metadata[:width] = upload.dimensions[:width]
+      blob.metadata[:height] = upload.dimensions[:height]
+      blob.metadata[:crop] = upload.crop_dimensions
+      blob.analyzed = true
+      blob.save
+      document.images.attach(blob)
+      redirect_to document_images_path(document), notice: "Image attached"
+    end
+  end
+
+private
+
+  def image_params
+    params.fetch(:document, {}).permit(:image)
+  end
+
+  class UploadedImageProcessorService
+    InvalidImage = Struct.new(:valid?, :errors)
+    ValidImage = Struct.new(
```

Comment:
It probably shouldn’t be possible for ValidImage.valid? to be false ;-)

Reply:
good point!